src/ports/postgres/modules/utilities/pivot.py_in

```diff
@@ -58,66 +61,255 @@ def pivot(schema_madlib, source_table, out_table,
-    # The aggregate collects pivot_values values for a given pivot_cols value
-    case_str = ("{agg}("
-                "CASE WHEN \"{{pcol}}\" = '{{value}}' THEN {pval} ELSE NULL END"
-                ")".
-                format(agg=aggregate_func,
-                       pval=pval_no_quotes))
-    sql_list.append(
-        ", " +
-        # Assign the name of the new column
-        ', '.join("{case_str} as \"{{pcol}}_{{value}}\"".
-                  format(case_str=case_str).
-                  format(pcol=pcol_no_quotes, value=str(value))
-                  for value in distinct_values if value is not None))
+    # Note that the distinct values are appended in order of users list
+    # This ordering is important when we access pivot_comb entries
+    distinct_str = ["SELECT "]
+    distinct_str.append(
+        ', '.join("array_agg(DISTINCT {pcol}) AS {pcol}_values".
+            format(pcol=pcol) for pcol in pcols))
+    distinct_str.append(" FROM " + source_table)
+    distinct_values = plpy.execute(''.join(distinct_str))
+
+    # Collect the distinc values for every pivot column
+    pcol_distinct_values = {}
+    pcol_max_length = 0
+    for pcol in pcols:
+        # Read the distinct values for this pcol
+        pcol_tmp = [item for item in distinct_values[0][pcol+"_values"]]
+        # Remove null values if keep null is not true
+        if not keep_null:
+            pcol_tmp = [x for x in pcol_tmp if x is not None]
+        pcol_distinct_values[pcol]=pcol_tmp
+        # Max length of the string that pcol values can create +
+        # length of pcol + 1 (for _ character)
+        pcol_max_length+=max([len(str(item)) for item in pcol_tmp])+len(pcol)+1
+
+    # For non postgres systems add null to the pcol_distinct_values if necessary
+    # This is required because DISTINCT keyword is implemented differently
+    is_pg = m4_ifdef(<!__POSTGRESQL__!>, True, False)
+    if keep_null and not is_pg:
+        null_str_begin = ["SELECT "]
+        null_str_begin.append(
+            ', '.join("array_agg({pcol}_isn) as {pcol}_isnull".
+                format(pcol=pcol) for pcol in pcols))
+        null_str_end = [" FROM (SELECT DISTINCT "]
+        null_str_end.append(
+            ', '.join(" (CASE WHEN {pcol} IS NULL THEN 1 END) "
+                " AS {pcol}_isn".
+                format(pcol=pcol) for pcol in pcols))
+        null_str_end.append(" FROM " + source_table +") x")
+        null_str = ''.join(null_str_begin) + ''.join(null_str_end)
+        null_values = plpy.execute(null_str)
+        for pcol in pcols:
+            pcol_tmp = [item for item in null_values[0][pcol+"_isnull"]]
+            if 1 in pcol_tmp:
+                pcol_distinct_values[pcol].append(None)
+
+    # Create the combination of every possible pivot column
+    # Assume piv and piv2 are pivot columns. piv=(1,2) and piv2=(3,4,5)
+    # pivot_comb = ((1,3),(1,4),(1,5),(2,3),(2,4),(2,5))
+    pivot_comb = list(itertools.product(*([pcol_distinct_values[pcol]
```

Comment:
I suggest keeping the product output as an iterator (i.e. don't make it a list) since you're only looping through it once. Further keep a count of number of pivot_comb seen while iterating and return a warning when you've reached the threshold. Finally, make the threshold a constant and initialize it at top with explanation on why it's set to that particular value (1000).